nbqa/handle_magics.py

```diff
@@ -210,4 +172,323 @@ class CellMagicHandler(MagicHandler):
+            re.sub(r"%\w+", lambda m: " " * len(m.group()), self._ipython_magic)
+        ).strip()
+        syntax_tree = self._get_syntax_tree(code)
+        if syntax_tree and self._contains_callable(syntax_tree):
+            self._contains_code = True
+            # remove the trailing backslashes if any from the code
+            code = re.sub(r"\\\n", "\n", code)
+            return self._MAGIC_TEMPLATE_WITH_CODE.format(
+                indented_code=indent(code, self._LEADING_SPACES), token=self._token
+            )
+
+        return super().replace_magic()
+
+    def restore_magic(self, cell_source: str) -> str:
+        """Return the cell source with the ipython magic restored.
+
+        Parameters
+        ----------
+        cell_source : str
+            Source code of the notebook cell
+
+        Returns
+        -------
+        str
+            Cell source with ipython magic restored.
+        """
+        if self._contains_code:
+            pattern = MagicHandler._get_regex_pattern(
+                self._MAGIC_WITH_CODE_REGEX_TEMPLATE, self._token
+            )
+            return self._restore_magic_with_modified_code(pattern, cell_source)
+
+        return super().restore_magic(cell_source)
+
+    def _restore_magic_with_modified_code(
+        self, replacement_code_pattern: Pattern[str], cell_source: str
+    ) -> str:
+        """Return cell source code with ipython magic statement restored.
+
+        Parameters
+        ----------
+        replacement_code_pattern : Pattern[str]
+            Regex pattern to find the python code replacing the ipython magic
+        cell_source : str
+            Source code of the notebook cell
+
+        Returns
+        -------
+        str
+            Cell source with ipython magic restored
+
+        Raises
+        ------
+        AssertionError
+            Raised when unable to extract the python code replacing ipython magic. This
+            error is handled with internally and not thrown to the caller.
+        """
+        ipython_magic = self._ipython_magic
+
+        # In case of any assertion failures, we should gracefully handle it
+        # by replacing the temporary python code with the original ipython magic
+        # Hence AssertionError is caught and ignored.
+        with contextlib.suppress(AssertionError):
+            # Get the code used for replacing the ipython magic statement
+            match_obj = replacement_code_pattern.search(cell_source)
+            if match_obj is None:  # pragma: nocover
+                raise AssertionError("unable to extract the replaced python code")
+
+            extracted_code = self._extract_code(match_obj.group())
+
+            # IPython magic is recreated from modified source so that formatters
+            # like black won't reformat the notebook every time they are run.
+            ipython_magic = f"""
+            {self._extract_line_magic()} {self._indent_extracted_code(extracted_code)}
+            """.strip()
```

Comment:
can the second line here be indented?

Reply:
This looks better. I didn't know this will concatenate two strings.